packages/devtools_app/test/network_controller_test.dart

```diff
@@ -0,0 +1,215 @@
+// Copyright 2020 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+@TestOn('vm')
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:devtools_app/src/globals.dart';
+import 'package:devtools_app/src/network/http_request_data.dart';
+import 'package:devtools_app/src/network/network_controller.dart';
+import 'package:devtools_app/src/service_manager.dart';
+import 'package:test/test.dart';
+import 'package:vm_service/vm_service.dart';
+
+import 'support/mocks.dart';
+import 'support/utils.dart';
+
+void main() {
+  group(
+    'NetworkController',
+    () {
+      NetworkController controller;
+      FakeServiceManager fakeServiceManager;
+      Timeline timeline;
+
+      setUpAll(
+        () async {
+          const testDataPath =
+              '../devtools_testing/lib/support/http_request_timeline_test_data.json';
+          final httpTestData = jsonDecode(
+            await File(testDataPath).readAsString(),
+          );
+          timeline = Timeline.parse(httpTestData);
+        },
+      );
+
+      setUp(
+        () {
+          fakeServiceManager = FakeServiceManager(
+            useFakeService: true,
+            timelineData: timeline,
+          );
+          setGlobal(ServiceConnectionManager, fakeServiceManager);
+          controller = NetworkController();
+        },
+      );
+
+      test(
+        'initialize recording state',
+        () async {
+          expect(
+            controller.isPolling,
+            false,
+          );
+
+          // Fake service pretends HTTP timeline logging is always enabled.
+          await controller.initialize();
+          expect(
+            controller.isPolling,
+            true,
+          );
+        },
+      );
+
+      test(
+        'start and pause recording',
+        () async {
+          expect(
+            controller.isPolling,
+            false,
+          );
+          final notifier = controller.recordingNotifier;
+          await addListenerScope(
```

Comment:
Couldn't the test still pass if the listener was never called?

Reply:
Yes, good point. I've added a check in `addListenerScope` to ensure the listener is actually called.